ads_async/asyncio/client.py

```diff
@@ -684,6 +688,146 @@ async def prune_unknown_notifications(self):
+            buffer_size=buffer_size
+        )
+        projects = {}
+        for fn in current_config.zip.filelist:
+            if fn.filename.startswith("_Config/PLC/"):
+                project_name, _ = os.path.splitext(os.path.basename(fn.filename))
+                tpzip_filename = f"{project_name}.tpzip"
+                projects[tpzip_filename] = utils.InMemoryZipFile(
+                    await self.get_file(
+                        f"boot/CurrentConfig\\{tpzip_filename}", buffer_size=buffer_size
+                    )
+                )
+        return projects
+
+    async def get_file(
+        self,
+        filename: str,
+        encoding: str = constants.ADS_ASYNC_STRING_ENCODING,
+        buffer_size: int = 16384,
+    ) -> bytes:
+        """
+        Download a file from the PLC.
+
+        Parameters
+        ----------
+        filename : str
+            The filename, the start which should be either "boot/" or "target/".
+        directory : DownloadPath
+            The path to download from (/Hard Drive/TwinCAT/3.1/...)
+
+        Returns
+        -------
+        data : bytes
+            The contents of the file.
+
+        Raises
+        ------
+        RequestFailedError
+            With DEVICE_NOTFOUND if the file is not found
+        AdsAsyncException
+            If the download doesn't proceed as expected
+        """
+        try:
+            directory, filename = filename.split("/", 1)
+            path = constants.DownloadPath[directory.lower()]
+        except Exception:
+            raise ValueError(
+                f"The filename should start with boot/ or target/. " f"Got: {filename}"
+            )
+
+        if path in (constants.DownloadPath.boot,):
+            file_info = await self.get_file_stat(
+                filename, directory=path, encoding=encoding
+            )
+            self.log.info("Downloading %s: %s", filename, file_info)
+        else:
+            self.log.info("Downloading %s", filename)
+
+        request = self.circuit.request_file_download(
+            filename=filename.encode(encoding), directory=path
+        )
+        res = await self.write_and_read(request, port=AmsPort.R3_SYSSERV)
+        if res is None:
+            raise exceptions.AdsAsyncException(
+                "Unable to get a handle for downloading the file"
+            )
+
+        handle = ctypes.c_uint32.from_buffer_copy(res.data).value
+        read_request = self.circuit.read_file_contents(handle, buffer_size=buffer_size)
+        clear_request = self.circuit.clear_file_handle(handle)
+        buffer = []
+        try:
+            while True:
+                res = await self.write_and_read(read_request, port=AmsPort.R3_SYSSERV)
+                if res is None:
+                    raise exceptions.AdsAsyncException("Download interrupted?")
+                buffer.append(res.data)
+                if len(res.data) < buffer_size:
+                    break
```

Comment:
Is it at all possible for this loop to get stuck if the last bit of data fills the last buffer size exactly?

Reply:
It could fill the buffer exactly, but the next loop should get 0 bytes and then exit